### PR TITLE
csv: Implement CSV support library

### PR DESCRIPTION
**Pull request by tarruda**
This implements a CSV library that can be used as the basis of implementing CSV plugins/parsers on fluent-bit.

The library is designed with a simple re-entrant API where you feed bytes, and the parse function tells you if it needs more data to complete a full CSV record. 

It can optionally receive a callback that will be invoked whenever a CSV column is fully parsed (idea similar to a SAX parser). If no callback is passed, then the library can only be used for splitting CSV records (which may span across multiple lines).

A note about the tests: It may seem complex, but I've tested the same inputs multiple times, with each time having a different chunk size (from 1 to the full data). The reason for doing this is that we can't predict where CSV data will be split when reading from a file descriptor, so it is important to test every possible way the input can be fed.

Also did some fixes for the flb_file.c module, which was not properly creating a NUL terminated `flb_sds_t` instance.

### CODE REVIEW

File: include/fluent-bit/flb_csv.h [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+    size_t length;
+    size_t field_count;
+    int state;
+    bool field_parsed;
+    bool has_dquote;
+    void *data;
+};
+
+void flb_csv_init(
+        struct flb_csv_state *state,
```
Comment (edsiper) [inline include/fluent-bit/flb_csv.h:54]: please align parameters to the `(` level

File: src/flb_csv.c [hunk 2]
```diff
@@ -0,0 +1,316 @@
...
+    state->has_dquote = false;
+    state->field_parsed = false;
+    state->state = FLB_CSV_STATE_INITIAL;
+    flb_sds_len_set(state->escape_buffer, 0);
+    flb_sds_len_set(state->buffered_data, 0);
+    state->offset = 0;
+}
+
+static int invoke_field_callback(
+        struct flb_csv_state *state,
```
Comment (edsiper) [inline src/flb_csv.c:54]: please align parameters to the `(` level

File: src/flb_csv.c [hunk 3]
```diff
@@ -0,0 +1,316 @@
...
+        state->escape_buffer[escpos++] = buf[bufpos++];
+    }
+    state->escape_buffer[escpos] = 0;
+    flb_sds_len_set(state->escape_buffer, escpos);
+    state->field_callback(state->data, state->escape_buffer, escpos);
+    return 0;
+}
+
+static int parse_simple(
+        struct flb_csv_state *state,
```
Comment (edsiper) [inline src/flb_csv.c:93]: please align parameters to the `(` level

File: src/flb_csv.c [hunk 4]
```diff
@@ -0,0 +1,316 @@
...
+        }
+        state->offset++;
+    }
+
+    state->length = state->offset - state->start;
+    return 0;
+}
+
+static int parse_quoted(
+        struct flb_csv_state *state,
```
Comment (edsiper) [inline src/flb_csv.c:115]: please align parameters to the `(` level

File: src/flb_csv.c [hunk 5]
```diff
@@ -0,0 +1,316 @@
...
+        state->offset++;
+    }
+
+    /* subtract 1 to remove ending double quote */
+    state->length = state->offset - state->start - 1;
+    return 0;
+}
+
+static int parse_csv_field(
+        struct flb_csv_state *state,
```
Comment (edsiper) [inline src/flb_csv.c:146]: please align parameters to the `(` level

File: src/flb_csv.c [hunk 6]
```diff
@@ -0,0 +1,316 @@
...
+        if (ret) {
+            return ret;
+        }
+    }
+
+    return ret;
+}
+
+void flb_csv_init(
+        struct flb_csv_state *state,
```
Comment (edsiper) [inline src/flb_csv.c:210]: please align parameters to the `(` level

File: src/flb_csv.c [hunk 7]
```diff
@@ -0,0 +1,316 @@
...
+            return FLB_CSV_EOF;
+        }
+        c = **bufptr;
+        if (state->state == FLB_CSV_STATE_INITIAL) {
+            if (c == '\r') {
+                state->state = FLB_CSV_STATE_FOUND_CR;
+                (*bufptr)++;
+                (*buflen)--;
+                continue;
+            } else if (c == '\n') {
```
Comment (edsiper) [inline src/flb_csv.c:244]: `else if` must go in the next line: https://github.com/fluent/fluent-bit/blob/master/CONTRIBUTING.md#general-requirements

File: src/flb_csv.c [hunk 8]
```diff
@@ -0,0 +1,316 @@
...
+                /* accept single linefeed as record terminator, even
+                 * though the spec says to look for \r\n */
+                (*bufptr)++;
+                (*buflen)--;
+                break;
+            } else if (c == ',') {
```
Comment (edsiper) [inline src/flb_csv.c:250]: https://github.com/fluent/fluent-bit/blob/master/CONTRIBUTING.md#general-requirements

File: src/flb_csv.c [hunk 9]
```diff
@@ -0,0 +1,316 @@
...
+                    state->field_count++;
+                    if (state->field_callback) {
+                        /* empty field, but we need to invoke the callback anyway */
+                        state->field_callback(state->data, "", 0);
+                    }
+                }
+                state->field_parsed = false;
+                continue;
+            }
+        } else if (state->state == FLB_CSV_STATE_FOUND_CR) {
```
Comment (edsiper) [inline src/flb_csv.c:263]: https://github.com/fluent/fluent-bit/blob/master/CONTRIBUTING.md#general-requirements

**Review by edsiper — CHANGES_REQUESTED** [review]

**Comment by tarruda** [comment]
fixed the style errors

**Pull request merged**